### ISSUE TITLE
Passing in user sub as targetIdentityId to Amplify.Storage.uploadFile fails with 403 on iOS only

### ISSUE THREAD

**Issue by lfraker** [aws-amplify/amplify-flutter#720]
**Describe the bug**
I followed the process detailed here: https://github.com/aws-amplify/amplify-js/issues/54#issuecomment-876455852 to change the S3 key used in private files from Cognito Identity Id to User Pool sub.
When uploading a test private file to S3 using Amplify.Storage.uploadFile, with the targetIdentityId set to the user sub, I get the following error on an iOS Simulator:

```
StorageException (StorageException(message: The HTTP response status code is [403]., recoverySuggestion: TODO some status code to recovery message mapper.
For more information on HTTP status codes, take a look at
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes, underlyingException: null))
```

The same code works when running on an Android Emulator.

Here's the code:

```dart
    const String exampleString = 'Test file';
    final S3UploadFileOptions options = S3UploadFileOptions(
      accessLevel: StorageAccessLevel.private,
      targetIdentityId: userId,
    );
    final Directory tempDir = await getTemporaryDirectory();
    final File exampleFile = File(tempDir.path + '/test.txt')
      ..createSync()
      ..writeAsStringSync(exampleString);

    try {
      final UploadFileResult result = await Amplify.Storage.uploadFile(
        local: exampleFile,
        key: 'TestFileKey',
        options: options,
      );
      print('Successfully uploaded file: ${result.key}');
    } on StorageException catch (e) {
      print('Error uploading file: $e');
    }
```

**To Reproduce**
While debugging I tried rolling back the mapping in Cognito/S3 (so that IdentityId was being used again). It seemed like iOS behavior was still unexpected in the original configuration. This may be a simpler repro:
1. Set up a project with auth and storage
2. Configure the S3UploadFileOptions in the following way: 
```dart
  final S3UploadFileOptions options = S3UploadFileOptions(
        accessLevel: StorageAccessLevel.private,
        targetIdentityId: 'incorrect string',
  );
```
3. Try to execute an Amplify.Storage.uploadFile with the configured options.
4. Run on an Android emulator, and get an expected unauthorized error:
```
StorageException (StorageException(message: Something went wrong with your AWS S3 Storage upload file operation, recoverySuggestion: See attached exception for more information and suggestions, underlyingException: com.amazonaws.services.s3.model.AmazonS3Exception: Access Denied (Service: Amazon S3; Status Code: 403; Error Code: AccessDenied; Request ID: DCMKD8P0DDPDB26G), S3 Extended Request ID: RdXMlpn9CInv5nDRH6wUmguq+Usi4TWCwU2Tz1IaLB+JRWNwhw0DXjnQfTarlcrqTxoqMXj55WA=))
```
5. Run on iOS and get no error (the file is saved), even though an authorization error is expected with a fake targetIdentityId

Steps to reproduce the behavior exactly as detailed above:
1. Set up your Cognito Identity Pool and S3 instance to leverage user sub instead of IdentityId following the steps here: https://github.com/aws-amplify/amplify-js/issues/54#issuecomment-876455852
2. Try to execute Amplify.Storage.uploadFile with user sub configured in the passed in S3UploadFileOptions on an iOS simulator.
3. Get error:
```
StorageException (StorageException(message: The HTTP response status code is [403]., recoverySuggestion: TODO some status code to recovery message mapper.
For more information on HTTP status codes, take a look at
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes, underlyingException: null))
```

**Expected behavior**
In the simpler repro, I'd expect the app running on iOS to fail with the incorrect targetIdentityId. In the complete scenario, I'd expect the app running on iOS to succeed (as it does on Android) when the user sub is passed in as the targetIdentityId.

**Screenshots**

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[] Android
[x] iOS

<details>

  <summary>Output of <code>flutter doctor -v</code></summary>
  
  ```flutter doctor -v
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-x64, locale en-US)
    • Flutter version 2.2.3 at /Users/lukefraker/flutter
    • Framework revision f4abaa0735 (12 days ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/lukefraker/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2020.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 53.1.5
    • Dart plugin version 203.6912

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (2 available)
    • iPhone SE (2nd generation) (mobile) • E7A9FE65-4E77-4747-AF01-A5FA66D17DE3 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)                        • chrome                               • web-javascript • Google Chrome 91.0.4472.114

• No issues found!
```

</details>

**Smartphone (please complete the following information):**
 - Device: [iOS Simulator]
 - OS: [Mac OS]
 - Version [iOS 14.5]

**Additional context**
First time posting an issue, so please let me know if any info is missing or formatted incorrectly. 

For additional context:
I'm sharing private images uploaded to S3 between the list of owners on a Datastore model object. No one else should have access to these images. I'm using an function directive to generate a pre-signed url to the image when any of the owners loads the model object. As such, I need the image key to be construct-able with the user sub (what is populated in the owners' field) instead of Cognito IdentityId. If there's a better way to accomplish sharing a private S3 object between a small set of users, I am open to any suggestions?

The only workaround I can think of for this issue (untested) is to generate a 'putObject' pre-signed url with another function directive, and use this to write the file to S3 via filestream. If there's any better alternative while waiting for a resolution to this issue, also open to any suggestions here?

**Comment by Jordan-Nelson** [comment]
> I'm sharing private images uploaded to S3 between the list of owners on a Datastore model object. No one else should have access to these images. I'm using an function directive to generate a pre-signed url to the image when any of the owners loads the model object. As such, I need the image key to be construct-able with the user sub (what is populated in the owners' field) instead of Cognito IdentityId. If there's a better way to accomplish sharing a private S3 object between a small set of users, I am open to any suggestions?

Hi @lfraker - I want to make sure that I understand the use case. It sounds like you would like to restrict write access to the image to just the owner (user that uploaded the image) and then restrict read access to a specific group of users. Am I understanding that correctly?

**Comment by lfraker** [comment]
> As I was writing this, I thought of a variation on option one - haven't tested but curious if this would be a viable option. If I were to only upload/download via pre-signed URLs, could I generate a URL with a key that doesn't correspond to either user's sub (so no one user is the 'owner') and store the random sub id on the datastore object that both user's have access to. (e.g. photo key for 'getObject' and 'putObject' pre-signed URLs is 'private/<random stored id>/<photo key>')?

**Update:**

I was able to get this working. The lambda functions resolve to a getURL and putURL that either of the two users can use to view/update the photo. Outside of these pre-signed URLs no one can access the photo.

@Jordan-Nelson curious if you'd advise against the approach detailed above for any reason?

Additionally, there seems to be an issue where the URLs expire and the values in the on-device datastore become stale, requiring another sync from the backend (run the lambda functions again to generate fresh pre-signed URLs). Even with a longer expiry time, this could eventually occur (and I'd prefer to not max out expiry time). Clearing the local store resolves this, but I'd rather not run this every time a URL expires. Any pointers for triggering a selective sync (or a way to prevent caching a specific model type locally)?

Apologies as I'm getting away from the original issue. Happy to open a second issue if I can't self-resolve with the URL refresh. Regarding the original issue, at this time it doesn't seem like I'm still blocked on it, but feel free to keep it open if you want to investigate the targetIdentityId issue for iOS.

**Comment by Jordan-Nelson** [comment]
Hi @lfraker - For the original issue (403 when using user sub in place of identityId), this unfortunately is not a supported use case. There are work arounds (such as the one you linked to). We can mark this issue as a feature request for that if you would like, but there is also a feature request for this in open in the CLI repo - https://github.com/aws-amplify/amplify-cli/issues/1847. The issue in the CLI is for general support in amplify.

As you noted, generating a pre-signed url to upload the file is circumventing the built in security. There is added risk in generating a url that allows anyone with access to the url to overwrite the file, especially if that url is being shared, has a long expiration time, and/or is not immediately being expired after use.

The use case here (group based access in storage) is not officially supported. Any workaround will likely have risks/drawbacks, particularly related to security. If you would like I can mark this as a feature request to track interest in an officially supported way to accomplish group based access in storage.

**Comment by lfraker** [comment]
Hi @Jordan-Nelson thanks for the reply. for the user sub issue, I understand that the workaround is not the officially supported path. However, I'm curious what the point of the 'targetIdentityId'  property on 'S3UploadFileOptions' is then? As mentioned above, when running in Android (after performing the backend mapping to use user sub), I was able to pass in the user sub and correctly authorize. When running in iOS the storage libraries seem to completely ignore whatever value I passed in (garbage values were accepted before I performed the mapping and correct values were rejected after). Is this just a dummy field on iOS? Is this documented anywhere (the API docs are pretty light: https://pub.dev/documentation/amplify_storage_s3/latest/amplify_storage_plugin/S3UploadFileOptions-class.html)?
Regarding marking it as a feature request, I'm fine with it being tracked on the other thread. Hopefully that gets some traction as we're just about to hit its 2 year anniversary :).

For group-based access to storage, feel free to mark this as a feature request if you think it would help. I've come across a few similar issues: https://github.com/aws-amplify/amplify-cli/issues/3621 & https://github.com/aws-amplify/amplify-cli/issues/3515, and it seems like the resolutions have either been to use Cognito Groups if you fit within the restrictions (I do not), or go the pre-signed URL route (I understand that this is not an ideal solution from a security perspective, but there doesn't appear to be an alternative if I want to use Amplify).

**Comment by Jordan-Nelson** [comment]
The intended use of the `targetIdentityId` param is to pass in the user's Cognito identityId, which can be accessed via [amplify auth's fetchAuthSession](https://docs.amplify.aws/lib/auth/access_credentials/q/platform/flutter). Missing dart docs is a known issue at the moment. The [storage docs for access levels](https://docs.amplify.aws/lib/storage/configureaccess/q/platform/flutter) have a callout and link to the auth docs, but I agree that this could be improved (potentially with an example).

If you are seeing incorrect values being accepted on iOS (without the mapping), that sounds like an issue. If you are able to repro this, would you mind opening a separate issue for it?

**Comment by lfraker** [comment]
@Jordan-Nelson thanks for the details. I went back to the pre-mapping config since I am now generating keys which neither user can access without a pre-signed URL. With this config, I'm able to repro the original issue (storage upload succeeds on iOS even if you provide a garbage `targetIdentityId`. I'll log a separate issue for this.

Regarding this issue, if it would help to track it is a group-based storage access feature request, I can update the title. However, since this seems more like a core gap rather than an amplify-flutter request, if it doesn't make sense to track the request here, this issue can probably be closed.

**Comment by Jordan-Nelson** [comment]
I am closing this issue. For anyone looking to use the user sub instead of the identity ID, please refer to [this issue](https://github.com/aws-amplify/amplify-cli/issues/1847) in the CLI repo. Since this feature is not available in any amplify library, we will use the feature request in the CLI repo to track interest. Feel free to 👍 the issue to help track interest.
